File: app/src/test/java/com/kirchhoff/example/githubclient/presenters/AuthPresenterTest.java
package com.kirchhoff.example.githubclient.presenters;

import com.kirchhoff.example.githubclient.Constants;
import com.kirchhoff.example.githubclient.repository.FakeGitHubRepository;
import com.kirchhoff.example.githubclient.repository.FakeKeyValueStorage;
import com.kirchhoff.example.githubclient.repository.GitHubDataSource;
import com.kirchhoff.example.githubclient.repository.keyvalue.KeyValueStorage;
import com.kirchhoff.example.githubclient.ui.auth.AuthContract;
import com.kirchhoff.example.githubclient.ui.auth.AuthPresenter;
import com.kirchhoff.example.githubclient.utils.schedulers.BaseSchedulerProvider;
import com.kirchhoff.example.githubclient.utils.schedulers.ImmediateSchedulerProvider;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.MockitoAnnotations;

import static junit.framework.Assert.assertNotNull;

/**
 * @author Kirchhoff-
 */

@RunWith(JUnit4.class)
public class AuthPresenterTest {

    @Mock
    AuthContract.View view;

    private AuthPresenter presenter;

    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.initMocks(this);

        GitHubDataSource repository = new FakeGitHubRepository();
        BaseSchedulerProvider provider = new ImmediateSchedulerProvider();
        KeyValueStorage storage = new FakeKeyValueStorage();

        presenter = new AuthPresenter(repository, view, provider, storage);
    }


    @Test
    public void testCreated() throws Exception {
        assertNotNull(presenter);
        Mockito.verifyNoMoreInteractions(view);
    }

    @Test
    public void testEmptyLogin() throws Exception {
        presenter.auth("", Constants.PASSWORD);
        Mockito.verify(view).showLoginError();

        Mockito.verifyNoMoreInteractions(view);
    }

    @Test
    public void testEmptyPassword() throws Exception {
        presenter.auth(Constants.LOGIN, "");
        Mockito.verify(view).showPasswordError();

        Mockito.verifyNoMoreInteractions(view);
    }

    @Test
    public void testEmptyLoginAndPassword() throws Exception {
        presenter.auth("", "");
        Mockito.verify(view).showLoginError();
        Mockito.verify(view).showPasswordError();

        Mockito.verifyNoMoreInteractions(view);
    }

    @Test
    public void testAuthError() throws Exception {
        presenter.auth(Constants.WRONG_LOGIN, Constants.WRONG_PASSWORD);

        Mockito.verify(view).showLoading();
        Mockito.verify(view).hideLoading();
        Mockito.verify(view).showError();

        Mockito.verifyNoMoreInteractions(view);
    }

    @Test
    public void testSuccessAuth() throws Exception {
        presenter.auth(Constants.LOGIN, Constants.PASSWORD);

        Mockito.verify(view).showLoading();
        Mockito.verify(view).hideLoading();
        Mockito.verify(view).openRepositoryScreen();

        Mockito.verifyNoMoreInteractions(view);
    }
}

File: app/src/test/java/com/kirchhoff/example/githubclient/presenters/AuthPresenterTest.kt
package com.kirchhoff.example.githubclient.presenters

import com.kirchhoff.example.githubclient.Constants
import com.kirchhoff.example.githubclient.Injection
import com.kirchhoff.example.githubclient.ui.auth.AuthContract
import com.kirchhoff.example.githubclient.ui.auth.AuthPresenter
import junit.framework.Assert.assertNotNull
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import org.mockito.Mock
import org.mockito.Mockito.verify
import org.mockito.Mockito.verifyNoMoreInteractions
import org.mockito.MockitoAnnotations

/**
 * @author Kirchhoff-
 */
@RunWith(JUnit4::class)
class AuthPresenterTest {

    @Mock
    private lateinit var view: AuthContract.View

    private lateinit var presenter: AuthPresenter

    @Before
    @Throws(Exception::class)
    fun init() {
        MockitoAnnotations.initMocks(this)

        presenter = AuthPresenter(Injection.provideGitHubRepository(),
                view, Injection.provideSchedulerProvider(), Injection.provideKeyValueStorage())
    }

    @Test
    @Throws(Exception::class)
    fun testCreated() {
        assertNotNull(presenter)
        verifyNoMoreInteractions(view)
    }

    @Test
    @Throws(Exception::class)
    fun testEmptyLogin() {
        presenter.auth("", Constants.PASSWORD)
        verify(view).showLoginError()

        verifyNoMoreInteractions(view)
    }

    @Test
    @Throws(Exception::class)
    fun testEmptyPassword() {
        presenter.auth(Constants.LOGIN, "")
        verify(view).showPasswordError()

        verifyNoMoreInteractions(view)
    }

    @Test
    @Throws(Exception::class)
    fun testEmptyLoginAndPassword() {
        presenter.auth("", "")
        verify(view).showLoginError()
        verify(view).showPasswordError()

        verifyNoMoreInteractions(view)
    }

    @Test
    @Throws(Exception::class)
    fun testAuthError() {
        presenter.auth(Constants.WRONG_LOGIN, Constants.WRONG_PASSWORD)

        verify(view).showLoading()
        verify(view).hideLoading()
        verify(view).showError()

        verifyNoMoreInteractions(view)
    }

    @Test
    @Throws(Exception::class)
    fun testSuccessAuth() {
        presenter.auth(Constants.LOGIN, Constants.PASSWORD)

        verify(view).showLoading()
        verify(view).hideLoading()
        verify(view).openRepositoryScreen()

        verifyNoMoreInteractions(view)
    }
}
